cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/replication/MissingRevsReplicationTest.java

```diff
@@ -0,0 +1,93 @@
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+package com.cloudant.sync.replication;
+
+
+import static org.mockito.Matchers.anyInt;
+import static org.mockito.Matchers.anyMapOf;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.spy;
+
+import com.cloudant.mazha.ChangesResult;
+import com.cloudant.mazha.CouchClient;
+import com.cloudant.sync.datastore.DocumentRevision;
+import com.cloudant.sync.datastore.DocumentRevisionTree;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Matchers;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+
+public class MissingRevsReplicationTest extends ReplicationTestBase {
+
+    private CouchClient clientMock;
+
+    @Before
+    public void setupMocks() throws Exception {
+        // Partially mock the remote and replace it for our test
+        clientMock = spy(remoteDb.couchClient);
+        remoteDb.couchClient = clientMock;
+        couchClient = clientMock;
+    }
+
+    @Override
+    protected PullStrategy getPullStrategy() {
+        PullStrategy s = super.getPullStrategy();
+        s.sourceDb = remoteDb;
+        return s;
+    }
+
+    @Test
+    public void testReplicationWithMissingRevision() throws Exception {
+        // Create doc
+        Bar a = BarUtils.createBar(remoteDb, "testdoc", "alpha", 1);
+        // Update doc
+        final Bar b = BarUtils.updateBar(remoteDb, "testdoc", "beta", 2);
+
+        Bar c = BarUtils.updateBar(remoteDb, "testdoc", "gamma", 3);
+        // We have two remote revisions
+
+        doAnswer(new Answer<ChangesResult>() {
+
+            @Override
+            public ChangesResult answer(InvocationOnMock invocation) throws Throwable {
+                @SuppressWarnings("unchecked")
+                ChangesResult changesResult = (ChangesResult) invocation.callRealMethod();
+
+                // modify the changes feed so that it returns a revision that isn't a leaf revision.
+                changesResult.getResults().get(0).getChanges().get(0).setRev(b.getRevision());
+                return changesResult;
+
+            }
+        }).when(clientMock).changes(anyString(), anyMapOf(String.class, String.class), Matchers.anyObject(), anyInt());
+
+        // Do the pull replication
+        pull();
+        // Validate the document was created at the correct rev
+        DocumentRevision rev = datastore.getDocument("testdoc");
+        Assert.assertNotNull("The document should be present", rev);
+        Assert.assertEquals("The local revision should be the third generation",
+                c.getRevision(),
+                rev.getRevision());
+        // Validate that there is only a single revision
```

Comment:
This doesn't appear to be how the test works any more. A tree depth of 2 indicates 3 revisions.